c++/src/kj/string.h

```diff
@@ -289,6 +290,76 @@ String heapString(ArrayPtr<const char> value);
 // String.
 
 namespace _ {  // private
+template <typename T> struct IsSignalSafeToCharSequence_ {
+  static constexpr bool value =
+#if __cpp_char8_t
+    isSameType<T, const char8_t*>() ||
+#endif
+    isSameType<T, decltype(nullptr)>() || isSameType<Decay<T>, bool>() ||
+    isIntegral<Decay<T>>() || isSameType<Decay<T>, kj::StringPtr>() ||
+    isSameType<Decay<T>, kj::String>() || isSameType<T, const char *>() ||
+    isSameType<T, const void *>();
+};
+
+template <typename T>
+struct IsSignalSafeToCharSequence_<T&>: IsSignalSafeToCharSequence_<T> {};
+template <typename T>
+struct IsSignalSafeToCharSequence_<const T&>: IsSignalSafeToCharSequence_<T> {};
+template <typename T>
+struct IsSignalSafeToCharSequence_<T&&>: IsSignalSafeToCharSequence_<T> {};
+template <typename T>
+struct IsSignalSafeToCharSequence_<const T&&>: IsSignalSafeToCharSequence_<T> {};
+
+template <typename T>
+struct IsSignalSafeToCharSequence_<kj::Array<T>> {
+  static constexpr bool value = IsSignalSafeToCharSequence_<T>::value;
+};
+template <typename T>
+struct IsSignalSafeToCharSequence_<kj::ArrayPtr<T>> {
+  static constexpr bool value = IsSignalSafeToCharSequence_<T>::value;
+};
+template <size_t N>
+struct IsSignalSafeToCharSequence_<char (&)[N]> { static constexpr bool value = true; };
+template <size_t N>
+struct IsSignalSafeToCharSequence_<const char (&)[N]> { static constexpr bool value = true; };
+
+
+template <typename T, size_t N>
+struct IsSignalSafeToCharSequence_<CappedArray<T, N>> {
+  static constexpr bool value = IsSignalSafeToCharSequence_<T>::value;
+};
+template <typename T, size_t N>
+struct IsSignalSafeToCharSequence_<FixedArray<T, N>> {
+  static constexpr bool value = IsSignalSafeToCharSequence_<T>::value;
+};
+
+#define KJ_COMMA ,
+#define KJ_ANNOTATE_SIGNAL_SAFE_STRINGIFY(T, condition, ...)                   \
+  namespace kj { namespace _ {                                                 \
+  template <__VA_ARGS__> struct IsSignalSafeToCharSequence_<T> {               \
+    static constexpr bool value = (condition);                                 \
+  };                                                                           \
+  } }
+// Annotate that the provided type is safe to be converted to a string in a
+// signal handler context if some condition is met. T is the type you're
+// specializing. If it's a templated type with multiple arguments (i.e. it has
+// commas), you must use KJ_COMMA as parenthesis can't be made to work. The
+// variadic parts of the macro are additional parameters to templatize the first
+// argument on. For example, imagine some type
+//   template <typename Arg1, typename Arg2> struct MyType {...};
+// that we wish to annotate. You would do it this way:
+//   KJ_ANNOTATE_SIGNAL_SAFE_STRINGIFY(
+//       MyType<Arg1 KJ_COMMA Arg2>, <some condition>, typename Arg1, typename
```

Comment:
This KJ_COMMA thing is pretty gross.

What if we just said that signal-safe stringifiers should return `SignalSafe<T>`, which is a trivial struct wrapper around `T`? No need for a separate annotation macro at all.

If we really want a separate annotation macro, we should have it work the same way KJ_STRINGIFY works, so that the interface is like:

```
template <typename T, typename U>
KJ_STRINGIFY_IS_SIGNAL_SAFE(MyType<T, U>);

template <typename T, typename U>
KJ_STRINGIFY_CONDITIONALLY_SIGNAL_SAFE(AnotherType<T, U>) {
  return kj::isSignalSafeToCharSequence<T>():
}
```

However, I think returning `SignalSafe<T>` seems like a cleaner approach.

Reply:
I can investigate the `SignalSafe<T>` idea. It does seem more elegant & might simplify the detection stuff.